File: src/public/services/NotificationService.php
<?php
require_once __DIR__ . '/../models/Model.php';

class NotificationService {

    private $db;

    public function __construct() {
        $this->db = Model::getModel()->bd;
    }

    /**
     * Crée une notification pour un utilisateur
     */
    public function notifier(int $utilisateur_id, string $message, string $type = 'info') {
        $req = $this->db->prepare("
            INSERT INTO notification (utilisateur_id, message, type, date_creation, lu)
            VALUES (?, ?, ?, NOW(), 0)
        ");
        return $req->execute([$utilisateur_id, $message, $type]);
    }

    /**
     * Notifie le département qu'un colis est arrivé
     */
    public function notifierArriveColis(int $bon_commande_id) {
        $req = $this->db->prepare("
            SELECT u.id_utilisateur
            FROM utilisateur u
            JOIN bon_commande b ON u.departement_id = b.departement_id
            WHERE b.id_bon_commande = ?
        ");
        $req->execute([$bon_commande_id]);
        $utilisateurs = $req->fetchAll(PDO::FETCH_ASSOC);

        foreach ($utilisateurs as $u) {
            $this->notifier(
                $u['id_utilisateur'],
                "Un colis associé à votre bon de commande est arrivé.",
                'colis'
            );
        }
    }

    /**
     * Notifie le département que son devis a été validé ou refusé
     */
    public function notifierStatutDevis(int $devis_id, string $statut) {
        $req = $this->db->prepare("
            SELECT u.id_utilisateur
            FROM utilisateur u
            JOIN devis d ON u.departement_id = d.departement_id
            WHERE d.id_devis = ?
        ");
        $req->execute([$devis_id]);
        $utilisateurs = $req->fetchAll(PDO::FETCH_ASSOC);

        $message = $statut === 'valide'
            ? "Votre devis a été validé."
            : "Votre devis a été refusé.";

        foreach ($utilisateurs as $u) {
            $this->notifier($u['id_utilisateur'], $message, $statut);
        }
    }

    /**
     * Récupère les notifications non lues d'un utilisateur
     */
    public function getNotificationsNonLues(int $utilisateur_id): array {
        $req = $this->db->prepare("
            SELECT * FROM notification
            WHERE utilisateur_id = ? AND lu = 0
            ORDER BY date_creation DESC
        ");
        $req->execute([$utilisateur_id]);
        return $req->fetchAll(PDO::FETCH_ASSOC);
    }

    /**
     * Marque toutes les notifications comme lues
     */
    public function marquerToutesLues(int $utilisateur_id) {
        $req = $this->db->prepare("
            UPDATE notification SET lu = 1
            WHERE utilisateur_id = ?
        ");
        return $req->execute([$utilisateur_id]);
    }
}

File: python_app/services/notification_service.py
from config.database import get_db


class NotificationService:
    """Service de gestion des notifications utilisateurs."""

    def notifier(self, utilisateur_id: int, message: str, type_notif: str = 'info') -> bool:
        """Crée une notification pour un utilisateur."""
        try:
            with get_db().cursor() as c:
                c.execute(
                    """INSERT INTO notification (utilisateur_id, message, type, date_creation, lu)
                       VALUES (%s, %s, %s, NOW(), 0)""",
                    (utilisateur_id, message, type_notif)
                )
            return True
        except Exception:
            return False

    def notifier_arrive_colis(self, bon_commande_id: int):
        """Notifie le département qu'un colis associé à son bon de commande est arrivé."""
        with get_db().cursor() as c:
            c.execute(
                """SELECT u.id_utilisateur
                   FROM utilisateur u
                   JOIN bon_commande b ON u.departement_id = b.departement_id
                   WHERE b.id_bon_commande = %s""",
                (bon_commande_id,)
            )
            utilisateurs = c.fetchall()

        for u in utilisateurs:
            self.notifier(
                u['id_utilisateur'],
                "Un colis associé à votre bon de commande est arrivé.",
                'colis'
            )

    def notifier_statut_devis(self, devis_id: int, statut: str):
        """Notifie le département que son devis a été validé ou refusé."""
        with get_db().cursor() as c:
            c.execute(
                """SELECT u.id_utilisateur
                   FROM utilisateur u
                   JOIN devis d ON u.departement_id = d.departement_id
                   WHERE d.id_devis = %s""",
                (devis_id,)
            )
            utilisateurs = c.fetchall()

        message = "Votre devis a été validé." if statut == 'valide' else "Votre devis a été refusé."

        for u in utilisateurs:
            self.notifier(u['id_utilisateur'], message, statut)

    def get_notifications_non_lues(self, utilisateur_id: int) -> list:
        """Récupère les notifications non lues d'un utilisateur."""
        with get_db().cursor() as c:
            c.execute(
                """SELECT * FROM notification
                   WHERE utilisateur_id = %s AND lu = 0
                   ORDER BY date_creation DESC""",
                (utilisateur_id,)
            )
            return c.fetchall()

    def marquer_toutes_lues(self, utilisateur_id: int) -> bool:
        """Marque toutes les notifications d'un utilisateur comme lues."""
        try:
            with get_db().cursor() as c:
                c.execute(
                    "UPDATE notification SET lu = 1 WHERE utilisateur_id = %s",
                    (utilisateur_id,)
                )
            return True
        except Exception:
            return False
